content/blog/2019-10-23-global-value-numbering/index.md

```diff
@@ -0,0 +1,357 @@
+
+The second pass is a recursive renaming algorithm that uses a stack to determine the correct version of a variable name for every point in the program.
+The trickiest part of implementing this algorithm was traversing the basic blocks in the correct order.
+The high level algorithm specifies that you should recursively traverse the children of each block in the dominator tree. 
+However, we found that not only did we ned to traverse the children, but we needed to traverse them _in order_ relative to the control flow graph. 
+In particular, Bril’s ecosystem already had a utility for getting the dominators of every block. To get the dominance tree, we calculated the transitive reduction of this relation. 
+Our implementation additionally extending the dominance tree such that every list of children was ordered relative to a reverse post-order traversal of the control flow graph.
+
+In later testing (using an example, `test/gvn/constant-folding-tau-example.bril`, ported from the [class notes of Prof. Mooly Sagiv of Tel Aviv University)][tau], we found one minor bug with our SSA implementation. 
+Bril programs do not have a type checker and are aggressively dynamic: it’s perfectly valid to assign to a variable in one branch of a conditional but not the other. 
+The base SSA algorithm does not account for this, so we fail to properly rename in the second step of our algorithm. 
+For now, we work around this case by assigning a default value to to the variable before we branch. 
+
+[ssa]: https://en.wikipedia.org/wiki/Static_single_assignment_form
+[llvm]: https://llvm.org/docs/tutorial/OCamlLangImpl7.html
+[gcc]: https://gcc.gnu.org/onlinedocs/gccint/SSA.html
+[domf]: https://en.wikipedia.org/wiki/Dominator_(graph_theory)
+[tau]: http://www.cs.tau.ac.il/~msagiv/courses/pa07/lecture2-notes-update.pdf
+
+## Global value numbering
+
+There are two main approaches to global value numbering, as described in [Value Numbering][], Briggs, Cooper, and Simpson, 1997, both of which ultimately require input code in SSA form.
+Hash-based techniques, like those used in local value numbering, hash operations and find redundant values by comparing to previously hashed operations.
+Partitioning algorithms, on the other hand, divide all operations into equivalence classes.
+
+We chose to implement hash-based value numbering because it allows for copy propagation extensions and operates on the dominator tree computed during conversion to SSA.
+While converting to SSA and performing global value numbering can be performed together in one pass over the original code due to their similarity, we chose to implement them separately for modularity.
+A distinct SSA conversion allows more downstream optimizations to operate on the program's SSA form in addition to allowing testing of both passes independently.
+
+We implemented the following pseudocode, adapted from Figure 4 of the aforementioned paper, so called because it is a 'dominator-based value numbering technique'.
```

Comment:
```suggestion
We implemented the following pseudocode, adapted from Figure 4 of the aforementioned paper, so called because it is a "dominator-based value numbering technique."
```

In most :us: style guides, double quotes are the top-level thing, and periods go inside the punctuation.